doc/dev/error-categorization.md

```diff
@@ -0,0 +1,173 @@
+as the application just knows that it simply cannot access the desired resource
+and tells the user to grant it.
+
+Categories are leaf based, so you cannot put an error into a node (branch) such
+as `Permanent errors` (see below).
+
+If you feel for a new category, please forge a design decision document and make
+a PR to Elektra's repo.
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an
+error/warning in there.
+
+For a complete structural overview please visit the corresponding [design
+decision document](../decisions/error_codes.md)
+
+### Permanent errors ("C01000")
+
+The branch category `Permanent Errors` refer to such errors which cannot be
+fixed by retry at all. `Permanent Errors` are subdivided into
+
+- Resource
+- Installation
+- Logical
+
+#### Resource ("C01100")
+
+`Resource Errors` as a branch category are all kinds of errors which are either
+permission related, existence related or missing resources such as memory, RAM,
+etc. Examples are missing files, insufficient permissions or out of memory. This
+category forces users or applications to provide additional resources, create a
+missing file/ directory etc or to provide the relevant permissions.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `elektraMalloc`
+calls primarily as no more memory could be allocated for the application. Errors
+with not enough hard disc space do not belong here but into `General Resource`.
+Such errors will gain special handling in future releases and users cannot deal
+with such errors as of now.
+
+##### General Resource ("C01120")
+
+`General Resource Errors` are all kind of permission, existence and resource
+errors which are essential for Elektra to operate. Compared to validation errors
+this category has its focus the underlying system resources whereas validation
+errors for usages with specifications. Examples are missing files/ directories
+or insufficient permission to execute certain commands (eg. you would require
+sudo permissions). Reactions are fixing the permissions, remount, creating the
+file/directory and retry the operation. Compared to validation errors,
+administrators would change the specification or retry with a different value.
+
+#### Installation ("C01200")
+
+`Installation Errors` are those errors which are related to a wrong installation
+such as wrong plugin names, missing backends, initialization errors,
+misconfiguration of Elektra etc. Installation errors might also be non-Elektra
+specific but also from dependent library/applications such as gpg. Also plugin
+configuration errors belong to `Installation Errors` as this happens during
+mounting. Users will have to reconfigure, reinstall, recompile (with other
+settings) Elektra in order to get rid of this error or fix the installation of
+the corresponding library/application.
+
+#### Logical ("C01300")
```

Comment:
Maybe the top-level category is called "Programming". And "Logical" is the subcategory of bugs in Elektra.

Reply:
I thought about that but I do not think that "Programming" is a good self explanatory name. Also a introduction for another branch category does not yield a considerable benefit. I rewrote the logical section to not include "bugs" but only the subcategory. This also makes much more sense now.